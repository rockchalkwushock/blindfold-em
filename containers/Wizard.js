import { Component } from 'react'

import { Button, Frame } from '../components'
import Timer from './Timer'

class Wizard extends Component {
  state = {
    activity: '',
    cooldown: '',
    currentFrame: 1,
    timer: ''
  }
  _next = () => {
    let frame = this.state.currentFrame
    this.setState(state => ({ ...state, currentFrame: ++frame }))
  }
  _prev = () => {
    let frame = this.state.currentFrame
    this.setState(state => ({ ...state, currentFrame: --frame }))
  }
  handleOnChange = e => {
    const val = e.target.value
    const key = e.target.name
    this.setState(state => ({
      ...state,
      [key]: val
    }))
  }
  renderFrame = () => {
    const { activity, cooldown, currentFrame, timer } = this.state
    if (currentFrame === 1) {
      return (
        <Frame
          name="activity"
          fn={this.handleOnChange}
          text="What activity are you working on?"
          value={activity}
        />
      )
    } else if (currentFrame === 2) {
      return (
        <Frame
          name="timer"
          fn={this.handleOnChange}
          text="How long should the timer be?"
          value={timer}
        />
      )
    } else if (currentFrame === 3) {
      return (
        <Frame
          name="cooldown"
          fn={this.handleOnChange}
          text="How long should the cooldown be?"
          value={cooldown}
        />
      )
    }
  }
  renderTimer = () => (
    <Timer
      activity={this.state.activity}
      cooldown={this.state.cooldown}
      timer={this.state.timer}
    />
  )
  render() {
    const { currentFrame } = this.state
    const view = currentFrame !== 4 ? this.renderFrame() : this.renderTimer()
    return (
      <div>
        {view}
        <Button
          className="next"
          condition={currentFrame === 4}
          fn={this._next}
          text="Next"
        />
        <Button
          className="prev"
          condition={currentFrame === 1}
          fn={this._prev}
          text="Prev"
        />
      </div>
    )
  }
}

export default Wizard
