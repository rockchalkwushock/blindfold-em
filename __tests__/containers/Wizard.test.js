import React from 'react'
import { mount } from 'enzyme'

import Wizard from '../../containers/Wizard'

const initialState = {
  activity: '',
  cooldown: '',
  currentFrame: 1,
  timer: ''
}

const updateState = (state, newState) => ({
  ...state,
  ...newState
})

describe('Container: <Wizard />', () => {
  let wrapper
  let input
  let next
  let prev
  beforeEach(() => {
    wrapper = mount(<Wizard />)
    input = wrapper.find('input')
    next = wrapper.find('button').first()
    prev = wrapper.find('button').last()
  })

  describe('State Management', () => {
    test('1. should find initial state', () => {
      expect(wrapper.state()).toEqual(initialState)
    })
    test('2. should find updated state onChange event for `activity`', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      expect(wrapper.state()).toEqual(
        updateState(initialState, { activity: 'coding' })
      )
    })
    test('3. should find updated state onClick event to move to Frame 2', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      expect(wrapper.state()).toEqual(
        updateState(initialState, { activity: 'coding', currentFrame: 2 })
      )
    })
    test('4. should find updated state onChange event for `timer`', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          currentFrame: 2,
          timer: '12'
        })
      )
    })
    test('5. should find updated state onClick event to move to Frame 3', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      next.simulate('click')
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          currentFrame: 3,
          timer: '12'
        })
      )
    })
    test('6. should find updated state onChange event for `cooldown`', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      next.simulate('click')
      input.simulate('change', { target: { name: 'cooldown', value: '3' } })
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          cooldown: '3',
          currentFrame: 3,
          timer: '12'
        })
      )
    })
    test('7. should find updated state onClick event to move to Frame 4', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      next.simulate('click')
      input.simulate('change', { target: { name: 'cooldown', value: '3' } })
      next.simulate('click')
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          cooldown: '3',
          currentFrame: 4,
          timer: '12'
        })
      )
    })
    test('8. should find updated state onClick event to move back to Frame 3', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      next.simulate('click')
      input.simulate('change', { target: { name: 'cooldown', value: '3' } })
      next.simulate('click')
      prev.simulate('click')
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          cooldown: '3',
          currentFrame: 3,
          timer: '12'
        })
      )
    })
    test('9. should find updated state onChange event for second time on `cooldown`', () => {
      input.simulate('change', {
        target: { name: 'activity', value: 'coding' }
      })
      next.simulate('click')
      input.simulate('change', { target: { name: 'timer', value: '12' } })
      next.simulate('click')
      input.simulate('change', { target: { name: 'cooldown', value: '3' } })
      next.simulate('click')
      prev.simulate('click')
      input.simulate('change', { target: { name: 'cooldown', value: '100' } })
      expect(wrapper.state()).toEqual(
        updateState(initialState, {
          activity: 'coding',
          cooldown: '100',
          currentFrame: 3,
          timer: '12'
        })
      )
    })
  })

  describe('Button States', () => {
    test('should see "Prev" button as disabled', () => {
      expect(prev.props().disabled).toEqual(true)
    })
    test('should see both buttons as enabled', () => {
      wrapper.setState({ currentFrame: 2 })
      next = wrapper.find('button').first()
      prev = wrapper.find('button').last()
      expect(next.props().disabled).toEqual(false)
      expect(prev.props().disabled).toEqual(false)
    })
    test('should see "Next" button as disabled', () => {
      wrapper.setState({ currentFrame: 4 })
      next = wrapper.find('button').first()
      expect(next.props().disabled).toEqual(true)
    })
  })

  describe('Snapshots', () => {
    test('should render without exploding on Frame 1', () => {
      expect(wrapper).toMatchSnapshot()
    })
    test('should render without exploding on Frame 2', () => {
      expect(wrapper.setState({ currentFrame: 2 })).toMatchSnapshot()
    })
    test('should render without exploding on Frame 3', () => {
      expect(wrapper.setState({ currentFrame: 3 })).toMatchSnapshot()
    })
    test('should render without exploding on Timer', () => {
      expect(wrapper.setState({ currentFrame: 4 })).toMatchSnapshot()
    })
  })
})
