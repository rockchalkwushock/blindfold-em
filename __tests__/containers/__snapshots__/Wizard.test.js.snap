// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: <Wizard /> Snapshots should render without exploding on Frame 1 1`] = `
<Wizard>
  <div>
    <Frame
      fn={[Function]}
      name="activity"
      text="What activity are you working on?"
      value=""
    >
      <div>
        <SubTitle
          text="What activity are you working on?"
        >
          <h2>
            What activity are you working on?
          </h2>
        </SubTitle>
        <Input
          fn={[Function]}
          name="activity"
          value=""
        >
          <div>
            <input
              name="activity"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </Input>
      </div>
    </Frame>
    <Button
      condition={false}
      fn={[Function]}
      text="Next"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Next
      </button>
    </Button>
    <Button
      condition={true}
      fn={[Function]}
      text="Prev"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        Prev
      </button>
    </Button>
  </div>
</Wizard>
`;

exports[`Container: <Wizard /> Snapshots should render without exploding on Frame 2 1`] = `
<Wizard>
  <div>
    <Frame
      fn={[Function]}
      name="timer"
      text="How long should the timer be?"
      value=""
    >
      <div>
        <SubTitle
          text="How long should the timer be?"
        >
          <h2>
            How long should the timer be?
          </h2>
        </SubTitle>
        <Input
          fn={[Function]}
          name="timer"
          value=""
        >
          <div>
            <input
              name="timer"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </Input>
      </div>
    </Frame>
    <Button
      condition={false}
      fn={[Function]}
      text="Next"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Next
      </button>
    </Button>
    <Button
      condition={false}
      fn={[Function]}
      text="Prev"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Prev
      </button>
    </Button>
  </div>
</Wizard>
`;

exports[`Container: <Wizard /> Snapshots should render without exploding on Frame 3 1`] = `
<Wizard>
  <div>
    <Frame
      fn={[Function]}
      name="cooldown"
      text="How long should the cooldown be?"
      value=""
    >
      <div>
        <SubTitle
          text="How long should the cooldown be?"
        >
          <h2>
            How long should the cooldown be?
          </h2>
        </SubTitle>
        <Input
          fn={[Function]}
          name="cooldown"
          value=""
        >
          <div>
            <input
              name="cooldown"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </Input>
      </div>
    </Frame>
    <Button
      condition={false}
      fn={[Function]}
      text="Next"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Next
      </button>
    </Button>
    <Button
      condition={false}
      fn={[Function]}
      text="Prev"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Prev
      </button>
    </Button>
  </div>
</Wizard>
`;

exports[`Container: <Wizard /> Snapshots should render without exploding on Timer 1`] = `
<Wizard>
  <div>
    <h1>
      Timer View
    </h1>
    <Button
      condition={true}
      fn={[Function]}
      text="Next"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        Next
      </button>
    </Button>
    <Button
      condition={false}
      fn={[Function]}
      text="Prev"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        Prev
      </button>
    </Button>
  </div>
</Wizard>
`;
